opm/autodiff/BlackoilModelEbos.hpp

```diff
@@ -203,14 +203,26 @@ namespace Opm {
         /// \param[in] timer                  simulation timer
         /// \param[in, out] reservoir_state   reservoir state variables
         /// \param[in, out] well_state        well state variables
-        void prepareStep(const SimulatorTimerInterface& /*timer*/,
+        void prepareStep(const SimulatorTimerInterface& timer,
                          const ReservoirState& /*reservoir_state*/,
                          const WellState& /* well_state */)
         {
             if ( wellModel().wellCollection()->havingVREPGroups() ) {
                 updateRateConverter();
             }
 
+            // update the solution variables in ebos
+
+            // if the last time step failed we need to update the curent solution
+            // and recalculate the Intesive Quantities.
+            if ( timer.lastStepFailed() ) {
```

Comment:
To see if I get this: if lastStepFailed(), that means we have chopped the step-size and are trying again, we are setting the current solution equal to the start-of-timestep solution (which should be identical to that of the failed attempt) as an initial guess. Otherwise, we are at an entirely new timestep, and we must update the initial solution.

If that is correct, my remaining questions are:
 - Why do we invalidate the cache on line 220? Is it not automatically invalidated when assigning to the solution in 219?
 - Why do we *not* invalidate the "timeindex 1" cache? Is no caching used for that (I guess that is likely, as it is immutable during a timestep).
 - Are the solution({0,1}) objects storing automatic derivatives or not? If so, is there some flattening going on at some point?

Reply:
> Is it not automatically invalidated when assigning to the solution in 219?

no: when the solution changes, the IQ cache needs to be explicitly invalidated.

> Why do we not invalidate the "timeindex 1" cache? Is no caching used for that (I guess that is likely, as it is immutable during a timestep).

correct. time index 1 is not cached (at least if the storage term for that time index gets cached.) also note that the solution for time index 1 does not change here, so there is no need to invalidate that cache.  

> Are the solution({0,1}) objects storing automatic derivatives or not? If so, is there some flattening going on at some point?

solution objects are of type `Dune::BlockVector<PrimaryVariables>`. basically that's identical to `std::vector<std::array<Scalar, numEq>>` with additional handling for some quirks plus some amenities.

Reply:
> no: when the solution changes, the IQ cache needs to be explicitly invalidated.

That is something we could fix, though. Easiest way is to make mutable access through a new member like mutableSolution() or setSolution(), and make that set the dirty flag. (Note that we probably must use a different function name, otherwise calling solution() from a mutable context would set the dirty flag even if only reading from it.)


> correct. time index 1 is not cached (at least if the storage term for that time index gets cached.) also note that the solution for time index 1 does not change here, so there is no need to invalidate that cache.

It changes in the else branch which is the one I was thinking about, but it's ok that it is not cached. It seems a bit fragile than we need to know such details at this point in the code. Automatic cache invalidation as described above would help I think.